File: src/upiUtils.js
const QRCode = require('qrcode');

// Regex for standard UPI VPA validation
const UPI_REGEX = /^[a-zA-Z0-9.\-_]{2,256}@[a-zA-Z]{2,64}$/;

/**
 * Validate standard UPI VPA format
 */
function isValidUPI(upiId) {
  if (!upiId || typeof upiId !== 'string') return false;
  return UPI_REGEX.test(upiId.trim());
}

/**
 * Format amount to 2 decimal places or clean integer format
 */
function formatAmount(amount) {
  const num = parseFloat(String(amount).replace(/[^0-9.]/g, ''));
  if (isNaN(num) || num <= 0) return null;
  // Return formatted float (e.g., 300 or 300.50)
  return Number.isInteger(num) ? num.toString() : num.toFixed(2);
}

/**
 * Build standard UPI payment URL (NPCI standard)
 * upi://pay?pa=VPA&pn=NAME&am=AMOUNT&cu=INR&tn=NOTE
 */
function buildUPIUrl(upiId, amount, name = '', note = '') {
  const cleanUpi = upiId.trim();
  const cleanAmount = formatAmount(amount);
  
  if (!cleanAmount) {
    throw new Error('Invalid amount specified');
  }

  const params = new URLSearchParams();
  params.append('pa', cleanUpi);
  
  if (name && name.trim()) {
    params.append('pn', name.trim());
  } else {
    params.append('pn', 'Payee');
  }

  params.append('am', cleanAmount);
  params.append('cu', 'INR');

  if (note && note.trim()) {
    params.append('tn', note.trim());
  } else {
    params.append('tn', `Payment of Rs ${cleanAmount}`);
  }

  return `upi://pay?${params.toString()}`;
}

/**
 * Generate PNG QR Code Buffer (for Telegram sending)
 */
async function generateQRBuffer(upiUrl, options = {}) {
  const defaultOpts = {
    errorCorrectionLevel: 'H',
    type: 'png',
    margin: 2,
    width: 512,
    color: {
      dark: '#0f172a',  // sleek dark navy dots
      light: '#ffffff' // pure white background
    }
  };

  return await QRCode.toBuffer(upiUrl, { ...defaultOpts, ...options });
}

/**
 * Generate Base64 Data URL (for Web UI rendering)
 */
async function generateQRDataURL(upiUrl, options = {}) {
  const defaultOpts = {
    errorCorrectionLevel: 'H',
    margin: 2,
    width: 400,
    color: {
      dark: '#0f172a',
      light: '#ffffff'
    }
  };

  return await QRCode.toDataURL(upiUrl, { ...defaultOpts, ...options });
}

/**
 * Deep link generators for specific UPI apps
 */
function getAppLinks(upiUrl) {
  const encoded = encodeURIComponent(upiUrl);
  return {
    generic: upiUrl,
    phonepe: `phonepe://pay?${upiUrl.replace('upi://pay?', '')}`,
    gpay: `gpay://upi/pay?${upiUrl.replace('upi://pay?', '')}`,
    paytm: `paytmmp://pay?${upiUrl.replace('upi://pay?', '')}`
  };
}

module.exports = {
  isValidUPI,
  formatAmount,
  buildUPIUrl,
  generateQRBuffer,
  generateQRDataURL,
  getAppLinks
};

File: upi_utils.py
import re
import io
import base64
from urllib.parse import urlencode
import qrcode

UPI_REGEX = re.compile(r'^[a-zA-Z0-9.\-_]{2,256}@[a-zA-Z]{2,64}$')

def is_valid_upi(upi_id: str) -> bool:
    if not upi_id or not isinstance(upi_id, str):
        return False
    return bool(UPI_REGEX.match(upi_id.strip()))

def format_amount(amount) -> str:
    try:
        clean_str = re.sub(r'[^0-9.]', '', str(amount))
        val = float(clean_str)
        if val <= 0:
            return None
        if val.is_integer():
            return str(int(val))
        return f"{val:.2f}"
    except (ValueError, TypeError):
        return None

def build_upi_url(upi_id: str, amount, name: str = "", note: str = "") -> str:
    clean_upi = upi_id.strip()
    clean_amt = format_amount(amount)
    if not clean_amt:
        raise ValueError("Invalid amount specified")

    params = {
        'pa': clean_upi,
        'pn': name.strip() if name and name.strip() else 'Payee',
        'am': clean_amt,
        'cu': 'INR',
        'tn': note.strip() if note and note.strip() else f"Payment of Rs {clean_amt}"
    }

    return f"upi://pay?{urlencode(params, safe='@')}"

def generate_qr_bytes(upi_url: str) -> bytes:
    qr = qrcode.QRCode(
        version=1,
        error_correction=qrcode.constants.ERROR_CORRECT_H,
        box_size=10,
        border=2,
    )
    qr.add_data(upi_url)
    qr.make(fit=True)
    img = qr.make_image(fill_color="#0f172a", back_color="white")
    
    buffer = io.BytesIO()
    img.save(buffer, format="PNG")
    buffer.seek(0)
    return buffer.getvalue()

def generate_qr_data_url(upi_url: str) -> str:
    png_bytes = generate_qr_bytes(upi_url)
    base64_encoded = base64.b64encode(png_bytes).decode('utf-8')
    return f"data:image/png;base64,{base64_encoded}"

def get_app_links(upi_url: str) -> dict:
    raw_params = upi_url.replace("upi://pay?", "")
    return {
        "generic": upi_url,
        "phonepe": f"phonepe://pay?{raw_params}",
        "gpay": f"gpay://upi/pay?{raw_params}",
        "paytm": f"paytmmp://pay?{raw_params}"
    }
